### PR TITLE
support Dict(;sizehint) and use it to reduce allocations in Dict(itr)

### PR DESCRIPTION
**Pull request by LilithHafner**
Also make the same changes in WeakKeyDict.

The runtime improvement may be measurement noise; the allocation reduction is real.
```julia
julia> @btime Dict(i^5 => i for i in 1:60); # Before
  1.621 μs (7 allocations: 2.80 KiB)
julia> @btime Dict(i^5 => i for i in 1:60); # After
  1.480 μs (4 allocations: 2.39 KiB)
```

### CODE REVIEW

**Comment by KristofferC** [comment]
Is there a difference between this and just using `sizehint!`?

**Comment by simeonschaub** [comment]
IIUC, this avoids preallocating a vector for n=16 first that then needs to be resized anyways. Eventually, we might want to have a similar constructor for vectors as well since the current API with `sizehint!` doesn't seem ideal for use cases where you know the expected size in advance.

**Comment by LilithHafner** [comment]
> a similar constructor for vectors

For Vectors, we already have
```julia
v = Vector{T}(under, final_len)
v[1:n] .= initial_values
```
Which only allocates once, even if `final_len > length(initial_values)`. `Dict{K,V}(sizehint=n)` is analogous to `Vector{T}(under, n)`

**Comment by simeonschaub** [comment]
> > a similar constructor for vectors
> 
> For Vectors, we already have
> 
> ```julia
> v = Vector{T}(under, final_len)
> v[1:n] .= initial_values
> ```
> 
> Which only allocates once, even if `final_len > length(initial_values)`. `Dict{K,V}(sizehint=n)` is analogous to `Vector{T}(under, n)`

That isn't quite the same as `sizehint!` though. `sizehint!` is mostly meant for cases where you have a `push!`-based algorithm and might only have a rough estimate of the size required. There, `sizehint!` will still result in a length-0 vector at first, but Julia will try to overallocate enough memory that no resizing the buffer is required when pushing that number of elements to the vector.

**Comment by LilithHafner** [comment]
CI allocation checks were failing because of a performance bug in the previous `WeakKeyDict` constructor:

https://github.com/JuliaLang/julia/blob/a9e6b5adb50ec66db536dbc3bd88a0a20740e831/base/weakkeydict.jl#L16-L24
`Any` should be `WeakRef`

```julia
julia> @btime WeakKeyDict([i^5] => i for i in 1:100); # c9f139a, first commit of this PR
  22.599 μs (373 allocations: 21.39 KiB)

julia> @btime WeakKeyDict([i^5] => i for i in 1:100); # 94966e2, second commit of this PR
  18.857 μs (309 allocations: 14.20 KiB)
```

**Comment by LilithHafner** [comment]
I think CI failures are unrelated

File: base/dict.jl [hunk 1]
```diff
@@ -95,7 +98,7 @@ function Dict{K,V}(ps::Pair...) where V where K
     return h
 end
 # Note the constructors of WeakKeyDict mirror these here, keep in sync.
-Dict() = Dict{Any,Any}()
+Dict(;sizehint...) = Dict{Any,Any}(;sizehint...)
```
Comment (petvana) [inline base/dict.jl:101]:
```suggestion
Dict(;args...) = Dict{Any,Any}(;args...)
```
I know, there is only single argument right now, but `args` seems to be more commonly used for this case.

Comment (LilithHafner) [reply base/dict.jl:101]: I think this provides useful implicit documentation that is lost if we replace `sizehint...` with `args...` or [`kws...`](https://github.com/JuliaLang/julia/pull/46225)
```
julia> methods(Dict)
# 5 methods for type constructor:
[1] Dict(;sizehint...) in Base at dict.jl:118
[2] Dict(ps::Pair{K, V}...) where {K, V} in Base at dict.jl:124
[3] Dict(kv::Tuple{}) in Base at dict.jl:119
[4] Dict(ps::Pair...) in Base at dict.jl:125
[5] Dict(kv) in Base at dict.jl:127
```

Comment (fredrikekre) [reply base/dict.jl:101]:
```suggestion
Dict(; sizehint = 0) = Dict{Any,Any}(; sizehint)
```

File: base/weakkeydict.jl [hunk 2]
```diff
@@ -42,7 +42,7 @@ function WeakKeyDict{K,V}(ps::Pair...) where V where K
     end
     return h
 end
-WeakKeyDict() = WeakKeyDict{Any,Any}()
+WeakKeyDict(;sizehint...) = WeakKeyDict{Any,Any}(;sizehint...)
```
Comment (LilithHafner) [inline base/weakkeydict.jl:45]:
```suggestion
WeakKeyDict(; sizehint = 0) = WeakKeyDict{Any,Any}(; sizehint)
```

**Comment by KristofferC** [comment]
For now, maybe an internal `_dict_with_initial_size` could be made and used?

**Comment by LilithHafner** [comment]
This has gone stale and the performance win is small enough that I'm not interested in reviving it.

**Pull request closed without merging**